src/pretix/plugins/paypal/payment.py

```diff
@@ -549,136 +777,92 @@ def execute_refund(self, refund: OrderRefund):
-                for k, v in res.items():
-                    if k == 'sale':
-                        sale = paypalrestsdk.Sale.find(v['id'])
+            pp_payment = None
+            # Legacy PayPal - migrate the payment info_data first
+            if "purchase_units" not in refund.payment.info_data:
+                req = OrdersGetRequest(refund.payment.info_data['cart'])
+                response = self.client.execute(req)
+                refund.payment.info = json.dumps(response.result.dict())
+                refund.payment.save(update_fields=['info'])
+                refund.refresh_from_db()
+
+            for res in refund.payment.info_data['purchase_units'][0]['payments']['captures']:
+                if res['status'] in ['COMPLETED', 'PARTIALLY_REFUNDED']:
+                    pp_payment = res['id']
+                    break
+
+            if not pp_payment:
+                req = OrdersGetRequest(refund.payment.info_data['id'])
+                response = self.client.execute(req)
+                for res in response.result.purchase_units[0].payments.captures:
+                    if res['status'] in ['COMPLETED', 'PARTIALLY_REFUNDED']:
+                        pp_payment = res.id
                         break
 
-            if not sale:
-                pp_payment = paypalrestsdk.Payment.find(refund.payment.info_data['id'])
-                for res in pp_payment.transactions[0].related_resources:
-                    for k, v in res.to_dict().items():
-                        if k == 'sale':
-                            sale = paypalrestsdk.Sale.find(v['id'])
-                            break
-
-            pp_refund = sale.refund({
+            req = CapturesRefundRequest(pp_payment)
+            req.request_body({
                 "amount": {
-                    "total": self.format_price(refund.amount),
-                    "currency": refund.order.event.currency
+                    "value": self.format_price(refund.amount),
+                    "currency_code": refund.order.event.currency
                 }
             })
-        except paypalrestsdk.exceptions.ConnectionError as e:
+            response = self.client.execute(req)
+        except IOError as ioe:
             refund.order.log_action('pretix.event.order.refund.failed', {
                 'local_id': refund.local_id,
                 'provider': refund.provider,
-                'error': str(e)
+                'error': str(ioe)
             })
-            raise PaymentException(_('Refunding the amount via PayPal failed: {}').format(str(e)))
-        if not pp_refund.success():
+            raise PaymentException(_('Refunding the amount via PayPal failed: {}').format(str(ioe)))
+        if response.result.status == 'COMPLETED':
+            req = OrdersGetRequest(refund.payment.info_data['id'])
+            response = self.client.execute(req)
+            refund.payment.info = json.dumps(response.result.dict())
```

Comment:
This does not get saved. But also, it should not, refunding should not touch the historical orderpayment.